sdk/src/types/block/address/mod.rs

```diff
@@ -210,3 +232,240 @@ impl From<&Self> for Address {
+#[getset(get = "pub")]
+pub struct Restricted<A> {
+    address: A,
+    allowed_capabilities: Option<Capabilities>,
+}
+
+impl<A> Restricted<A> {
+    /// Creates a new [`Restricted`] address from the underlying type.
+    #[inline(always)]
+    pub fn new(address: A) -> Self {
+        Self {
+            address,
+            allowed_capabilities: Default::default(),
+        }
+    }
+
+    /// Sets the allowed capabilities flags.
+    #[inline(always)]
+    pub fn with_allowed_capabilities(mut self, allowed_capabilities: impl Into<Capabilities>) -> Self {
+        self.allowed_capabilities.replace(allowed_capabilities.into());
+        self
+    }
+
+    /// Sets the allowed capabilities flags.
+    #[inline(always)]
+    pub fn set_allowed_capabilities(&mut self, allowed_capabilities: impl Into<Capabilities>) -> &mut Self {
+        self.allowed_capabilities.replace(allowed_capabilities.into());
+        self
+    }
+}
+
+impl<A> From<A> for Restricted<A> {
+    fn from(value: A) -> Self {
+        Self::new(value)
+    }
+}
+
+impl<A: 'static + Packable> Packable for Restricted<A>
+where
+    Error: From<A::UnpackError>,
+{
+    type UnpackError = Error;
+    type UnpackVisitor = A::UnpackVisitor;
+
+    fn pack<P: packable::packer::Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        self.address.pack(packer)?;
+        if self
+            .allowed_capabilities
+            .map(|c| c.0 != CapabilityFlag::NONE)
+            .unwrap_or_default()
+        {
+            self.allowed_capabilities.pack(packer)?;
+        } else {
+            0_u8.pack(packer)?;
+        }
+        Ok(())
+    }
+
+    fn unpack<U: packable::unpacker::Unpacker, const VERIFY: bool>(
+        unpacker: &mut U,
+        visitor: &Self::UnpackVisitor,
+    ) -> Result<Self, packable::error::UnpackError<Self::UnpackError, U::Error>> {
+        let address = A::unpack::<_, VERIFY>(unpacker, visitor).coerce()?;
+        let allowed_capabilities_set = u8::unpack::<_, VERIFY>(unpacker, &()).coerce()? != 0;
+        let allowed_capabilities = allowed_capabilities_set
+            .then(|| Capabilities::unpack::<_, VERIFY>(unpacker, &()).coerce())
+            .transpose()?;
+        Ok(Self {
+            address,
+            allowed_capabilities,
+        })
+    }
+}
+
+pub struct CapabilityFlag;
+
+impl CapabilityFlag {
+    pub const NATIVE_TOKENS: u8 = 0b00000001;
+    pub const MANA: u8 = 0b00000010;
+    pub const TIMELOCKED_OUTPUTS: u8 = 0b00000100;
+    pub const EXPIRING_OUTPUTS: u8 = 0b00001000;
+    pub const STORAGE_DEPOSIT_OUTPUTS: u8 = 0b00010000;
+    pub const ACCOUNT_OUTPUTS: u8 = 0b00100000;
+    pub const NFT_OUTPUTS: u8 = 0b01000000;
+    pub const DELEGATION_OUTPUTS: u8 = 0b10000000;
+    pub const NONE: u8 = 0;
+    pub const ALL: u8 = u8::MAX;
+}
+
+#[derive(Copy, Clone, Debug, Default, Eq, PartialEq, Ord, PartialOrd, Hash, From, Deref)]
+#[repr(transparent)]
+pub struct Capabilities(u8);
```

Comment:
I don't have the TIP with me right now but I feel like I remember we could have more than one byte of capablities ?

Reply:
Not yet, and if we do that unfortunately it ruins a lot of code because `Address` is no longer `Copy`.

Reply:
Not yet but the protocol still plans for it and serialize a length, I'm not sure it's being done here

Reply:
It is, because option adds 0 or 1 and that's all that's currently allowed

Reply:
Ah, I didn't even know 0 was allowed. Hmm, still feels weird to not already plan for this but I'll think about it tomorrow.